src/lxc/conf.c

```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
 	}
 
 	if (flags & LXC_AUTO_SHMOUNTS_MASK) {
-		char *new_mount;
-		size_t len_mount;
+		size_t len_cont;
+		char *full_cont_path;
+		struct stat sb;
 		int ret = -1;
 
+		/* Construct the shmount path under the container root */
 		/* +1 for slash */
-		size_t len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
+		len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
 		/* +1 for the terminating '\0' */
-		char *full_cont_path = malloc(len_cont + 1);
+		full_cont_path = malloc(len_cont + 1);
 		if(!full_cont_path) {
 			SYSERROR("Not enough memory");
 			return -ENOMEM;
 		}
-
 		ret = snprintf(full_cont_path, len_cont + 1, "%s/%s", conf->rootfs.mount, conf->lxc_shmount.path_cont);
 		if (ret < 0 || ret >= len_cont + 1) {
 			SYSERROR("Failed to create filename");
-			goto out_free_path;
+			free(full_cont_path);
+			return -1;
 		}
 
+		/* Check if shmount point is already set up */
 		if(is_shared_mountpoint(conf->lxc_shmount.path_host)
-				&& is_shared_mountpoint(full_cont_path)) {
-			/* Already set up */
-			ret = 0;
-			goto out_free_path;
+				&& !stat(full_cont_path, &sb) && S_ISDIR(sb.st_mode)) {
+			/* The host path is a shared mounpoint
+			 * (TODO: not checked for MS_REC),
+			 * the container path exists --
+			 * so we only need to add it to mount_list
+			 * for the following setup */
+			goto finalize;
 		}
 
-		/* Borrowed from mount_entry_create_dir_file() */
+		/* Borrowed from mount_entry_create_dir_file()
+		 *
+		 * Create host and cont mount paths
+		 */
 		ret = mkdir_p(conf->lxc_shmount.path_host, 0755);
 		if (ret < 0 && errno != EEXIST) {
 			SYSERROR("Failed to create directory \"%s\"", conf->lxc_shmount.path_host);
-			goto out_free_path;
+			free(full_cont_path);
+			return ret;
+		}
+		ret = mkdir_p(full_cont_path, 0755);
+		if (ret < 0 && errno != EEXIST) {
+			SYSERROR("Failed to create directory \"%s\"", full_cont_path);
+			free(full_cont_path);
+			return ret;
 		}
 
+		/* Prepare host mountpoint */
 		ret = mount("tmpfs", conf->lxc_shmount.path_host, "tmpfs",
 					0, "size=100k,mode=0711");
 		if (ret < 0) {
 			SYSERROR("Failed to mount \"%s\"", conf->lxc_shmount.path_host);
-			goto out_free_path;
+			free(full_cont_path);
+			return ret;
 		}
-
 		ret = mount(conf->lxc_shmount.path_host, conf->lxc_shmount.path_host, "none",
 					MS_REC | MS_SHARED, "");
 		if (ret < 0) {
 			SYSERROR("Failed to make shared \"%s\"", conf->lxc_shmount.path_host);
-			goto out_free_path;
+			free(full_cont_path);
+			return ret;
 		}
 
-		/* +1 for the separating whitespace */
-		len_mount = strlen(conf->lxc_shmount.path_host) + 1 + len_cont + sizeof(" none bind,create=dir 0 0") - 1;
-		/* +1 for the termintating '\0' */
-		new_mount = malloc(len_mount + 1);
-		if (!new_mount) {
-			SYSERROR("Not enough memory");
-			goto out_free_path;
+finalize:
+		ret = add_shmount_to_conf(conf, full_cont_path);
```

Comment:
You shouldn't and can't add `full_cont_path` here. What you should simply add is the relative target for the bind-mount in the container. So if the bind-mount target for the shared mountpoint in the container is `/dev/.lxc` then you should pass in:

    add_shmount_to_conf(conf, "dev/.lxc")

since LXC will automatically take relative targets to be relative to the container's rootfs. :)